### PR TITLE
[SCFToCalyx] Support scf.execute_region operations [9/13]

### PR DESCRIPTION
**Pull request by mortbopet**
Inlines Calyx ExecuteRegionOp operations within their parent blocks. An execution region op (ERO) is inlined by:
 1. add a sink basic block for all yield operations inside the ERO to jump to
 2. Rewrite scf.yield calls inside the ERO to branch to the sink block
 3. inline the ERO region

### CODE REVIEW

**Review by mikeurbach — APPROVED** [review]
Looks good.

I still have the same concerns I raised on the initial PR: this is growing to support a number of different inputs, and I'm not sure what the use-cases are. What frontend generates ExecuteRegion ops? Do we have a concrete need for this support? If not, might be best to keep the pass simpler. I actually had the same concern about std.cond_br, by the way. If we are targetting the SCF level, shouldn't we be focused on scf.if instead?

I don't want to block progress, so if you want to keep landing the original patch in smaller pieces that is a-ok. I just think this might be a good point to stop and think about what inputs we need to support versus what inputs we could support. It's easy to try and boil the ocean, so having concrete use-cases coming from the frontend helps limit the scope. For example, what do [these](https://github.com/llvm/circt/blob/a3797fb83958fd21f48536db6e394a9204a24f9a/test/Analysis/dependence-analysis.mlir) looks like in SCF? Do we need ExecuteRegion (or cond_br for that matter)?

**Comment by mortbopet** [comment]
I'd hope to land the current commit series such that we have a baseline for trying out some end-to-end flows which goes through Calyx. Once we have some experience with the flow, what works, and what doesn't, we can start to generalize this pass. I'd like to avoid sinking in too much time getting a "perfect, general, ..." implementation up front, before we have some experience with exercising the tool.

In the future, i envision that this pass evolves into "StandardToCalyx" pass + the ability to register external dialect lowerings (just like what the LLVM exporter does). Through this, we can add external lowering logic for SCF ops, staticlogic pipeline, etc.

**Comment by mikeurbach** [comment]
I agree it's worth getting something working, so please do feel free to keep landing the original patch series. I'm just suggesting this specific patch might end up being dead code unless some end-to-end flow actually generates ExecuteRegion ops.

**Comment by mortbopet** [comment]
Hmm, that is true - it's existence is mainly based on the use of `scf.execute_region` in the `scf-for-to-while` pass. However, since that no longer relies on `scf.execute_region`, it indeed reduces the utility of this pass. I think i'll keep it here for now, since the code is there and tested...

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 1]
```diff
@@ -815,6 +815,54 @@ LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
...
+    auto *execOpEntryBlock = &execOp.getRegion().front();
+    auto *postBlock = execOp->getBlock()->splitBlock(execOp);
+    rewriter.inlineRegionBefore(execOp.getRegion(), postBlock);
+    rewriter.mergeBlocks(postBlock, preBlock);
+    rewriter.eraseOp(execOp);
+
+    /// Finally, erase the unused entry block of the execOp region.
+    rewriter.mergeBlocks(execOpEntryBlock, preBlock);
+
+    return success();
```
Comment (cgyurgyik) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:862]: Very clean. Nice job.

**Review by cgyurgyik — APPROVED** [review]
So right now we're in this weird phase of breaking down your mega-PR into smaller pieces. Therefore, we are assuming that these pieces will be used in future PRs. According to the dialogue above, it seems that `ExecuteRegionOp` will not necessarily be used?

> I think i'll keep it here for now, since the code is there and tested...

I disagree with this. If you see `ExecuteRegionOp` being used, then let's keep it. Otherwise, I don't think it should remain in the code base, given it is essentially dead code. If you're unsure, then please at least create an issue and add a `TODO(#<issue-number>)`. This will make future clean-up easier to maintain.

**Comment by mortbopet** [comment]
After doing a bit of digging, Polygeist might be a potential user for`scf.execute_region` lowering. An example is when building affine loops: https://github.com/wsmoses/Polygeist/blob/main/mlir-clang/Lib/clang-mlir.cc#L794.

I am planning on using Polygeist for work that comes after this commit series has landed. Given this I'll keep the code in for now, but make an issue to track the usefulness of the lowering of `scf.execute_region`. If it turns out that `scf.execute_region` are rarely or never generated, then i'd be ready to remove the code.

**Comment by cgyurgyik** [comment]
> https://github.com/wsmoses/Polygeist/blob/main/mlir-clang/Lib/clang-mlir.cc#L794

Perfect. Thanks @mortbopet !

**Pull request merged**